eax/eax.go

```diff
@@ -0,0 +1,174 @@
+// Copyright (C) 2019 ProtonTech AG
+
+// Package eax provides an implementation of the EAX
+// (encrypt-authenticate-translate) mode of operation, as described in
+// Bellare, Rogaway, and Wagner "THE EAX MODE OF OPERATION: A TWO-PASS
+// AUTHENTICATED-ENCRYPTION SCHEME OPTIMIZED FOR SIMPLICITY AND EFFICIENCY."
+// In FSE'04, volume 3017 of LNCS, 2004
+
+package eax
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"errors"
+	"golang.org/x/crypto/internal/byteutil"
+)
+
+const (
+	defaultTagSize   = 16
+	defaultNonceSize = 16
+)
+
+type eax struct {
+	block     cipher.Block // Only AES-{128, 192, 256} supported
+	tagSize   int          // At least 12 bytes recommended
+	nonceSize int
+}
+
+func (e *eax) NonceSize() int {
+	return e.nonceSize
+}
+
+func (e *eax) Overhead() int {
+	return e.tagSize
+}
+
+// NewEAX returns an EAX instance with AES-{keyLength} and default tag and
+// nonce lengths.
+func NewEAX(key []byte) (cipher.AEAD, error) {
```

Comment:
Probably it makes sense to pass a `block` instead of a `key` here too, like in gcm and ocb?

Reply:
The conflicting thing about this is that, sure, for similarity we could pass a block, but we are only supporting AES so we need to check for that. However, the `Block` interface from the `cipher` package does not export the name of the block cipher being implemented, and we cannot rely on the Stringer of AES either. In other words, I do not see a simple (and more importantly, reliable) way to test if AES has been passed here. I'd opt for security and initialize AES ourselves.

Reply:
Where/why do we require that the block cipher is AES?

Reply:
[It would be ok to do something like this,](https://stackoverflow.com/questions/39171678/instanceof-equivalent-in-golang) but the `aes` package does not define a type either.

Reply:
> Where/why do we require that the block cipher is AES?

Hmm, I suppose I was just mimicking [OpenPGP.js](https://openpgpjs.org/) on this, but now I see that there is reason to impose AES in js, and also that there is no such constraint on [RFC4880](https://tools.ietf.org/html/draft-ietf-openpgp-rfc4880bis-07#section-5.16.1). Therefore I agree with your suggestion